klib/cloudwatch.c

```diff
@@ -98,6 +135,305 @@ define_closure_function(1, 1, void, cw_aws_cred_handler,
+{
+    buffer auth = aws_req_sign(cw.h, cw.region, service, "POST", req, body,
+                               cw.access_key, cw.secret);
+    if (!auth)
+        return false;
+    set(req, sym(Authorization), auth);
+    set(req, sym(x-amz-security-token), alloca_wrap_cstring(cw.token));
+    status s = http_request(cw.h, out, HTTP_REQUEST_METHOD_POST, req, body);
+    boolean success = is_ok(s);
+    if (!success)
+        timm_dealloc(s);
+    deallocate_buffer(auth);
+    return success;
+}
+
+closure_function(2, 0, void, cw_send_async,
+                 void *, sender, ip_addr_t, addr)
+{
+    void (*sender)(const ip_addr_t *server) = bound(sender);
+    sender(ip_addr_isany(&bound(addr)) ? 0 : &bound(addr));
+    closure_finish();
+}
+
+static void cw_dns_cb(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    void (*sender)(const ip_addr_t *server) = callback_arg;
+    thunk t = closure(cw.h, cw_send_async, sender, ipaddr ? *ipaddr : ip_addr_any);
+    if (t != INVALID_ADDRESS)
+        async_apply(t);
+    else
+        msg_err("failed to allocate closure\n");
+}
+
+static void cw_connect(const char *server, void (*handler)(const ip_addr_t *server))
+{
+    ip_addr_t cw_host;
+    lwip_lock();
+    err_t err = dns_gethostbyname(server, &cw_host, cw_dns_cb, handler);
+    lwip_unlock();
+    switch (err) {
+    case ERR_OK:
+        cw_dns_cb(server, &cw_host, handler);
+        break;
+    case ERR_INPROGRESS:
+        break;
+    default:
+        cw_dns_cb(server, 0, handler);
+    }
+}
+
+static void cw_log_send_async(void)
+{
+    if ((cw.log_pending == 0) && (vector_length(cw.log_entries) > 0) &&
+        !timer_is_active(&cw.log_timer))
+        register_timer(kernel_timers, &cw.log_timer, CLOCK_ID_MONOTONIC, seconds(10), false, 0,
+                       (timer_handler)&cw.log_timer_handler);
+}
+
+closure_function(1, 1, void, cw_logstream_vh,
+                 boolean, parsed,
+                 value, v)
+{
+    bound(parsed) = true;
+    tuple resp = get_tuple(v, sym(start_line));
+    if (resp) {
+        buffer word;
+        for (u64 i = 0; (word = get(resp, intern_u64(i))); i++)
+            if (buffer_strstr(word, "OK") == 0)
+                goto success;
+    }
+    buffer content = get_string(v, sym(content));
+    if (content && (buffer_strstr(content, "ResourceAlreadyExistsException") >= 0))
+        goto success;
+    msg_err("unexpected response %v\n", v);
+    return;
+  success:
+    spin_lock(&cw.lock);
+    cw.log_inited = true;
+    cw_log_send_async();
+    spin_unlock(&cw.lock);
+}
+
+closure_function(3, 1, boolean, cw_log_meta_ibh,
+                 buffer_handler, parser, boolean *, parsed, buffer_handler, out,
+                 buffer, data)
+{
+    status s = apply(bound(parser), data);
+    if (!is_ok(s)) {
+        msg_err("failed to parse response: %v\n", s);
+        timm_dealloc(s);
+        if (data)
+            apply(bound(out), 0);
```

Comment:
Should this call closure_finish() if data is null here?

Reply:
Indeed we should call closure_finish() if data is null, I changed the code accordingly.